### PR TITLE
Diagonal brakes and block brakes

### PR DESCRIPTION
**Pull request by spacek531**
Speedy block brakes is merged 🎉 and this is the follow-up PR - implementing diagonal brakes and block brakes.

For paint code reasons, track elements with multiple sequence blocks now have all their sequence blocks updated when modifying brakes or brake speed.

### CODE REVIEW

**Comment by spacek531** [comment]
The build should be in a playable state now. There are a few outstanding things to do like fixing looping and LIM sprites. Waiting on #19922 and potentially another PR I have yet to make before this can be merged.

**Comment by karst** [comment]
Keep this open please.
Any progress on this?

File: src/openrct2/sprites.h
Comment (karst) [inline src/openrct2/sprites.h:1371]: Make sure these are listed under     `// G2 Track additions` with the appropriate track types.

**Comment by SpartanFrederic104** [comment]
Keep this open please

**Comment by spacek531** [comment]
I have given X123M3-256 write access, he has the motivation to finish the PR in a timely manner.

**Comment by X123M3-256** [comment]
Does anyone know why the CI is failing?

**Comment by duncanspumpkin** [comment]
its a replay test failure so something has diverged from how it was before. Are we expecting existing logic to have changed at all?

**Comment by X123M3-256** [comment]
I don't think so but I didn't create the PR so I'm not sure exactly what was changed (@spacek531?)

**Comment by fidwell** [comment]
I see this error message a lot in the build output: `Replay network version mismatch: '0.3.5.1-15', expected: '0.4.5-13'`. So maybe it just needs rebased / main merged in, since the original PR is way behind the current replay.

**Comment by X123M3-256** [comment]
> I see this error message a lot in the build output: `Replay network version mismatch: '0.3.5.1-15', expected: '0.4.5-13'`. So maybe it just needs rebased / main merged in, since the original PR is way behind the current replay.

I did rebase it, but I rebased it on top of my local develop which is a little out of date, so I can try rebasing again. The network version hasn't increased since I last pulled develop, though, it's still 0.4.5-12, so I think 0.4.5-13 is the correct network version because this PR requires it to be increased?

**Comment by X123M3-256** [comment]
I rebased it again on top of the latest develop, seems it's still failing.

**Comment by ZehMatt** [comment]
> I see this error message a lot in the build output: `Replay network version mismatch: '0.3.5.1-15', expected: '0.4.5-13'`. So maybe it just needs rebased / main merged in, since the original PR is way behind the current replay.

This is just the network version the replays were recorded with, this has nothing to do with the actual replay failing. The current one fails because something has changed that influences the outcome of the simulation.

**Comment by X123M3-256** [comment]
What is a replay? What do you mean recorded? It builds and works on my machine, I find it odd that some of the Linux builds succeed and others are failing?

**Comment by spacek531** [comment]
> its a replay test failure so something has diverged from how it was before. Are we expecting existing logic to have changed at all?

Off the top of my head, all tiles of multi-tile block brake tracks (diagonal chain up-25/60-to-flat, cable lift hill) will have their brake flag set instead of only tile 0. I don't remember if there are other changes.

**Comment by duncanspumpkin** [comment]
> What is a replay? What do you mean recorded? It builds and works on my machine, I find it odd that some of the Linux builds succeed and others are failing?

We record a 1000 ticks of a bpb save file and record the exact properties of every single entity in a replay. If you make a change to the logic that changes how an entity performs anything (assuming that action happens in bpb during the 1000ticks) then it will show up has a replay failure. It takes a little bit of time to run these replays so we don't run it on every ci just Windows 32/64, Mac, Linux gcc/clang. The idea is that it will catch accidental changes to logic that we weren't expecting. If we have on purpose made a change then i just rerecord the replay. I can also get the replay to generate a difference report that will say exactly which entity fields have changed but its not very reliable doing that and it often goes wrong (and it can only tell you the difference after 1000 ticks but the divergence might have happened much earlier and the butterfly effect might mess up lots).

**Comment by X123M3-256** [comment]
>  I can also get the replay to generate a difference report that will say exactly which entity fields have changed

How do I do this? I'd quite like to get this PR finished off because it's been almost 2 years since it was first implemented, but I've not encountered this issue before.

**Comment by spacek531** [comment]
> I can also get the replay to generate a difference report that will say exactly which entity fields have changed but its not very reliable doing that and it often goes wrong (and it can only tell you the difference after 1000 ticks but the divergence might have happened much earlier and the butterfly effect might mess up lots).

You are referring to the desync report, correct? When I used it previously it would identify the first frame(s) of the desync instead of the last.

My expectations for a good outcome on the desync report are that the desync consists of only TrackElement brake flags being set, and peep rand is the same on left and right.

**Comment by spacek531** [comment]
Replays are now working. Turns out I accidentally flipped a case when implementing the new TrackTypeIs* functions. 

As far as I am aware, these are the current outstanding items:

- [ ] Move sprites around in g2 according to @karst 's comment
- [ ] Add sprites for RMC coasters

**Comment by karst** [comment]
Spawning of vehicles on diagonal blockbrakes doesn't seem to quite work yet. When opening a ride it loops the sound 5 times and the actual vehicles dont spawn on the correct spot but rather somewhere seemingly random in the circuit.

**Comment by spacek531** [comment]
The issue with spawning trains is now fixed.

**Comment by spacek531** [comment]
I am marking this as ready for review. There are a few track types sprites have not yet been implemented for, but those can be added in a follow-up PR if they do not get added before the PR is accepted.

File: src/openrct2/network/NetworkBase.cpp
Comment (spacek531) [inline src/openrct2/network/NetworkBase.cpp:46]: Bump park version once the PR is reviewed.

**Comment by X123M3-256** [comment]
@spacek531 What would you say still needs doing on this? If it's the sprites, I would suggest removing any that you think aren't good enough in order to get the code itself in, because new sprites can always be added later. I'll try and get around to adding sprites for the hybrid and single rail tomorrow, I've already implemented support for diagonal brakes in my rendering code so I just need to render the sprites and test everything.

**Comment by spacek531** [comment]
It is ready for review. There are no known bugs. More sprites can be added while 0.4.6 is not yet out which will flesh out the feature more.

Currently missing track types:

- [ ] wooden coaster
- [ ] side friction coaster
- [ ] mini coaster
- [ ] giga coaster
- [ ] LIM launched coaster
- [ ] suspended swinging coaster

**Comment by spacek531** [comment]
Did you add the new image files? CI has the error
```
Could not import image file: D:\a\OpenRCT2\OpenRCT2\resources\g2\track\rmc\brake_diag_1.png
```

**Comment by karst** [comment]
Thisone will be requiring review now

File: src/openrct2/ride/TrackData.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:703]: Make this match the style of the others next to this.

File: src/openrct2/ride/TrackData.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:4037]: Should have TrackElemType:: in comments

File: src/openrct2/ride/TrackData.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:4570]: Same here

File: src/openrct2/ride/Track.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:567]: Don't make this a member function and rename it so that its clear this is a multi tile set brake.

File: src/openrct2/ride/VehicleSubpositionData.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/VehicleSubpositionData.cpp:30079]: TrackElemType::

File: src/openrct2/ride/coaster/HybridCoaster.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/coaster/HybridCoaster.cpp:1297]: ? Why this change. Has the order of sprites changed or something?

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
See minor comments. I've only glanced over the paint code I assume that is mostly a copy paste job?

File: src/openrct2/ride/coaster/HybridCoaster.cpp
Comment (X123M3-256) [reply src/openrct2/ride/coaster/HybridCoaster.cpp:1297]: There were redundant sprites. Flat pieces have rotational symmetry so you only need two angles, not four (and vanilla tracks only have two sprites for these pieces), but for some reason I forgot this and had rendered four sprites for the diagonal flat pieces.

I didn't want to make the brake sprites similarly redundant, and I didn't want it to be inconsistent, so I removed the duplicates.

File: src/openrct2/ride/Track.cpp
Comment (spacek531) [reply src/openrct2/ride/Track.cpp:567]: What would be the preference? Passing TrackElement as an argument?

Comment (duncanspumpkin) [reply src/openrct2/ride/Track.cpp:567]: Yeah just pass track element by ref as an arg. The main reason I'm not wanting it a member function is that its iterating onto other elements when on multi tile elements. I would expect member functions to only be touching the singular tile.

File: src/openrct2/ride/TrackData.cpp
Comment (spacek531) [reply src/openrct2/ride/TrackData.cpp:703]: The way I did it is the style that TrackSequenceProperties originally used. The style with the track name after was introduced by X7123M3-256 for his track pieces. Is that the way you want them all to be?

Comment (duncanspumpkin) [reply src/openrct2/ride/TrackData.cpp:703]: Argh... hmm keep as is then.

**Comment by spacek531** [comment]
Yes, the paint code is copy and pasted between all rides except for the height of the bboxes.

File: src/openrct2/ride/Ride.h
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.h:1112]: Could you go for a bit more descriptive name. SetBrakeMultiTileClosed would make it clear that this sets all the track elements in the track block. I know it's also doing single tile but that's just an optimisation.

File: src/openrct2/ride/VehicleSubpositionData.cpp
Comment (spacek531) [reply src/openrct2/ride/VehicleSubpositionData.cpp:30079]: that's only the diagonal turn pieces, the rest have no prefix

**Review by duncanspumpkin — APPROVED** [review]

**Comment by spacek531** [comment]
@X123M3-256 can you re-push the hybrid and single rail commits? I lost them while rebasing.

**Comment by spacek531** [comment]
does anyone know why linux bionic is erroring?

```
../src/openrct2/ride/coaster/WoodenRollerCoaster.cpp:8893:5: error: 'constexpr BoundBoxXYZ::BoundBoxXYZ()' used before its definition
```

**Comment by karst** [comment]
it seems the include `../../paint/Boundbox.h` is not located where it used to be.

**Comment by karst** [comment]
@duncanspumpkin Can you take a look at thisone one more time after all these small fixes?
I believe the only thing that has changed is the addition of a few more sprites + paintcode. @spacek531 can you verify?

**Comment by spacek531** [comment]
yes

**Review by karst — APPROVED** [review]
I have done one more playtest checking the functionality and it seems identical to its normal counter part. Very happy with this implementation, thx for the hard work :)

**Comment by Realsteel89** [comment]
A krimi this branch but clean work @Spacek I will test it again tomorrow.

**Comment by spacek531** [comment]
I don't have any more planned changes for this branch. The ride types that do not have brakes will be covered in a folllow-up PR.

**Review by duncanspumpkin — APPROVED** [review]
Minor issue with park version but other than that good to go.

**Pull request merged**
